Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,467 @@
+    var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    private enum RequiredAction {
+        case make(EventLoopPromise<ConnectionProvider>)
+        case existing(ConnectionProvider)
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> RequiredAction {
+        return self.lock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .existing(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                self._connectionProviders[key] = .future(promise.futureResult)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
+                    case .failure:
+                        self[key] = nil
+                    }
+                }
+                return .make(promise)
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key: Key
+        // Created a key and checks the URL components are valid
+        do {
+            key = try Key(url: request.url)
```

Comment:
why not pass request itself? I think it already has all parameters resolved and verified

Reply:
@artemredkin Do you mean making a `init(request: HTTPClient.Request)` for `Key`?
While some values are already resolved, others such as the port aren't, so we would still have to call `request.url.port` for instance.
Isn't the url also already validated when passed here?

Reply:
I just saw the request also contains a `port` property, in this case I'll change this to take Request as an argument. Thank you

Reply:
Fixed!